specs/plasma.md

```diff
@@ -0,0 +1,208 @@
+- [Security Considerations](#security-considerations)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+Using [Plasma][vitalikblog] mode means switching to an offchain data availability provider.
+The protocol guarantees availability with a challenge contract on L1 where users
+can challenge the availability of input data used to derive the chain.
+Challenging an input commitment means forcing providers to submit the input data on chain
+to prove the data is available during a time period. We call the period during which any party
+can challenge a commitment `challenge_window` and the period during which any party
+can submit the input data to resolve the challenge `resolve_window`.
+If a challenge is not resolved by the time `resolve_window` of L1 blocks have passed,
+chain derivation must be reset, omitting the input data when rederiving therefore causing a reorg.
+
+[vitalikblog]: https://vitalik.eth.limo/general/2023/11/14/neoplasma.html
+
+## DA Storage
+
+Input data is uploaded to the storage layer via plain HTTP calls to the DA storage service.
+This service is horizontally scalable and concerned with replicating the data across prefered storage layers
+such as IPFS or any S3 compatible storage. Input data is content addressed by its hash in the request url
+so responses are easily cachable.
+
+Any DA provider can implement the following endpoints to receive and serve input data:
+
+- ```text
+  Request:
+    POST /put/<hex_encoded_commitment>
+    Content-Type: application/octet-stream
+    Body: <preimage_bytes>
+  
+  Response:
+    200 OK
+  ```
+
+- ```text
+  Request:
+    GET /get/<hex_encoded_commitment>
+  
+  Response:
+    200 OK
+    Content-Type: application/octet-stream
+    Body: <preimage_bytes>
+  ```
+
+## Input Commitment
+
+The batching and compression of input data remain unchanged. When a batch is ready
+to be submitted to the inbox address, the data is uploaded to the DA storage layer instead, and a
+commitment (keccak256 hash) is submitted as the bacher inbox transaction call data.
+The batcher SHOULD NOT submit a commitment onchain unless input data was successfully stored on the service.
+In addition, a DA provider storage service SHOULD return an error response if it is unable to properly
+store the request payload so as to signal to the batcher to retry.
+Input commitments submitted onchain without proper storage on the DA provider service are subject to
+challenges if the input cannot be retrieved during the challenge window, as detailed in the following section.
+
+## Data Availability Challenge Contract
+
+### Parameters
+
+| Variable                      | Description                                                                 |
+| ----------------------------- | --------------------------------------------------------------------------- |
+| challengeWindow:  uint256 | L1 blocks period a commitment MAY be challenged after it's included onchain |
```

Comment:
I think number of L1 blocks makes more sense here, but in some places we use L1 blocks and in other places we use seconds. Slot number maps exactly to time, but I don't think that block number does.